docs/cookbook/text_translation.md

```diff
@@ -0,0 +1,350 @@
+        with print_progress_message(current_count):
+            evaluation = await evaluate_translation(original_text, translation_text)
+
+            improved_translation = await improve_translation(
+                translation_text=translation_text,
+                original_text=translation_text,
+                tone=tone,
+                audience=audience,
+                evaluation=evaluation,
+            )
+            translation_text = improved_translation.content
+    return translation_text
+
+
+asyncio.run(
+    multi_pass_translation(
+        text,
+        tone=Tone.casual,
+        audience=Audience.general,
+        pass_count=3,
+    )
+)
+```
+
+This technique allows for gradual improvement in various aspects such as grammar, vocabulary, and style, resulting in more natural and accurate translations.
+
+### 3. Multi-Provider Translation
+
+Multi-Provider translation involves using multiple LLM providers in parallel and comparing their results.
+
+```python
+import asyncio
+from collections.abc import Callable
+from contextlib import contextmanager
+from enum import StrEnum
+
+from pydantic import BaseModel, Field
+
+from mirascope.core import anthropic, gemini, openai, prompt_template
+
+
+def create_call_function(
+    call, model, _prompt_template, response_model=None
+) -> Callable:
+    return call(model, response_model=response_model)(_prompt_template)
+
+
+@contextmanager
+def print_progress_message(model: str, count: int):
+    print(f"{model=} Multi-pass translation start times: {count}")
+    yield
+    print(f"{model=} Multi-pass translation end times: {count}")
+
+
+async def multi_pass_translation(
+    original_text: str,
+    tone: Tone,
+    audience: Audience,
+    pass_count: int,
+    parametrized_translate: Callable,
+    evaluate_translation: Callable,
+    improve_translation: Callable,
+    model: str,
+) -> str:
+    with print_progress_message(model, 1):
+        translation = await parametrized_translate(
+            original_text, tone=tone, audience=audience
+        )
+        translation_text = translation.content
+
+    for current_count in range(2, pass_count + 1):
+        with print_progress_message(model, current_count):
+            evaluation = await evaluate_translation(original_text, translation_text)
+
+            improved_translation = await improve_translation(
+                translation_text=translation_text,
+                original_text=translation_text,
+                tone=tone,
+                audience=audience,
+                evaluation=evaluation,
+            )
+            translation_text = improved_translation.content
+    return translation_text
+
+
+async def multi_provider_translation(
+    original_text: str,
+    tone: Tone,
+    audience: Audience,
+    pass_count: int,
+    call_models: list[tuple[Callable, str]],
+) -> None:
+    results = []
+    for call, model in call_models:
+        parametrized_translate = create_call_function(
```

Comment:
these refactored `create_call_function` methods are too refactored I think. I would rather see either (1) `BasePrompt` definitions where you use the `run_async` method or (2) functions with no call decorator that have the decorator applied in this function.

Personally, I not the biggest fan of option (2) and would lean more toward option (1) using `BasePrompt` (because the methods without a call decorator are weird methods that don't do anything without the decorator).

Note the following are just examples and will need to be properly applied here.

## Option 1: `BasePrompt`

```python
@prompt_template(...)
class ParametrizedTranslatePrompt(BasePrompt):
    text: str
    tone: Tone
    audience: Audience

    async def translate(self, call: Callable, model: str) -> str:
        response = await self.run_async(call(model))
        return response.content


prompt = ParametrizedTranslatePrompt(...)
translation = asyncio.run(prompt.translate(call, model))
```

Doing it this way means you can also push some of the other stuff like `Evaluation` response model inside the class, e.g.:

```python
@prompt_template(...)
class EvaluateTranslationPrompt(BasePrompt):
    original_text: str
    translation_text: str

    async def evaluate(self, call: Callable, model: str) -> Evaluation:
        return await self.run_async(call(model, response_model=Evaluation))
        

prompt = EvaluateTranslationPrompt(...)
evaluation = prompt.evaluate(call, model)
```

## Option 2: Function

```python
@prompt_template(...)
async def parametrized_translate(text: str, tone: Tone, audience: Audience) -> None: ...

...

provider_specific_parameterized_translate = call(model)(parametrized_translate)
```

You would then do the same for all of the other methods.